### DISCUSSION TITLE
What is the difference between ISpecimenBuilder and ICustomization

### DISCUSSION THREAD

**Discussion by alban-ameti** [AutoFixture/AutoFixture#1449] [Q&A] [answered]
I would like to know what is the difference between a Customization and SpecimenBuilder and an example when one would be used instead of the other.

To me they both seem to do the same thing? I am not sure how they differ any help would be greatly appreciated.

**Comment by aivascu** [comment, marked as answer]
@alban-ameti the interfaces have very different meaning in AutoFixture. To better understand what they're doing, you need to first understand the way the Fixture object works. You can read the wiki article, about the [internal architecture](https://github.com/AutoFixture/AutoFixture/wiki/Internal-Architecture).

In short, when creating an object of a certain type, the fixture will break it down into its composing pieces (members/dependencies). Each piece will run through the pipeline as a distinct request. The process will run recursively until the initial request is broken down into requests for primitive values, that the fixture knows how to instantiate. The instantiation of these primitive request, is usually delegated to objects implementing `ISpecimenBuilder`. These are the most atomic part of the fixture pipeline.
Once the fixture starts replying with concrete objects, the replies are being assembled into object graphs.

The `ICustomization` objects usually configure the pipeline, by introducing or removing `ISpecimenBuilder` objects from the fixture. 
The most common customization points are, the `IFixture.Customizations` (the front of the pipeline), and `IFixtuer.ResidueCollectors` (the end of the pipeline).

Obviously this is an oversimplification of the process. If you wish to know more, see the resources specified in the [Documentation](https://github.com/AutoFixture/AutoFixture?tab=readme-ov-file#documentation) section, of the readme.

As a user you'll usually implement `ISpecimenBuilder` when you need to reply to a certain request with a custom response, like custom built object, and you'll implement `ICustomization` to encapsulate adding/removing specimen builders from the fixture, for certain scenarios.

**Reply by alban-ameti** [reply]
Thank you for that. That makes sense.
